proptest/src/strategy/lazy.rs

```diff
@@ -0,0 +1,161 @@
+//-
+// Copyright 2017 Jason Lingle
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use core::mem;
+use crate::std_facade::{Arc, fmt};
+
+use crate::strategy::traits::*;
+use crate::test_runner::*;
+
+/// Represents a value tree that is initialized on the first call to any
+/// methods.
+///
+/// This is used to defer potentially expensive generation to shrinking time.
+pub(crate) struct LazyValueTree<S : Strategy> {
+    state: LazyValueTreeState<S>,
+}
+
+enum LazyValueTreeState<S : Strategy> {
+    Initialized(S::Tree),
+    Uninitialized {
+        strategy: Arc<S>,
+        runner: TestRunner,
+    },
+    Failed,
+}
+
+impl<S : Strategy> LazyValueTree<S> {
+    /// Create a new value tree where initial generation is deferred until
+    /// `maybe_init` is called.
+    pub(crate) fn new(strategy: Arc<S>, runner: &mut TestRunner) -> Self {
+        let runner = runner.partial_clone();
+        Self {
+            state: LazyValueTreeState::Uninitialized { strategy, runner },
+        }
+    }
+
+    /// Create a new value tree that has already been initialized.
+    pub(crate) fn new_initialized(value_tree: S::Tree) -> Self {
+        Self { state: LazyValueTreeState::Initialized(value_tree) }
+    }
+
+    /// Returns a reference to the inner value tree if initialized.
+    pub(crate) fn as_inner(&self) -> Option<&S::Tree> {
+        match &self.state {
+            LazyValueTreeState::Initialized(v) => Some(v),
+            LazyValueTreeState::Uninitialized { .. }
+            | LazyValueTreeState::Failed => None,
```

Comment:
Unless you want to intentionally guard against additions to `LazyValueTreeState` you can just use `_ => None`.
(Applies throughout in this file)

Reply:
Yeah :) I tend to not use _ because I've run into issues with additions to enums in the past.